frontend/src/metabase/query_builder/components/QuestionActions.tsx

```diff
@@ -0,0 +1,226 @@
+
+import {
+  checkDatabaseSupportsModels,
+  checkCanBeModel,
+  checkDatabaseCanPersistDatasets,
+} from "metabase/lib/data-modeling/utils";
+
+import DatasetMetadataStrengthIndicator from "./view/sidebars/DatasetManagementSection/DatasetMetadataStrengthIndicator/DatasetMetadataStrengthIndicator";
+
+import { PLUGIN_MODERATION } from "metabase/plugins";
+
+import {
+  QuestionActionsContainer,
+  PopoverContainer,
+} from "./QuestionActions.styled";
+
+import { MODAL_TYPES } from "metabase/query_builder/constants";
+
+const ICON_SIZE = 18;
+
+const ADD_TO_DASH_TESTID = "add-to-dashboard-button";
+const MOVE_TESTID = "move-button";
+const TURN_INTO_DATASET_TESTID = "turn-into-dataset";
+const TOGGLE_MODEL_PERSISTENCE_TESTID = "toggle-persistence";
+const CLONE_TESTID = "clone-button";
+const ARCHIVE_TESTID = "archive-button";
+
+interface Props {
+  isBookmarked: boolean;
+  handleBookmark: () => void;
+  onOpenModal: (modalType: string) => void;
+  question: any;
+  setQueryBuilderMode: (
+    mode: string,
+    opt: { datasetEditorTab: string },
+  ) => void;
+  turnDatasetIntoQuestion: () => void;
+}
+
+const QuestionActions = ({
+  isBookmarked,
+  handleBookmark,
+  onOpenModal,
+  question,
+  setQueryBuilderMode,
+  turnDatasetIntoQuestion,
+}: Props) => {
+  const bookmarkButtonColor = isBookmarked ? color("brand") : "";
+  const bookmarkTooltip = isBookmarked ? t`Remove from bookmarks` : t`Bookmark`;
+
+  const isDataset = question.isDataset();
+  const canWrite = question.canWrite();
+
+  return (
+    <QuestionActionsContainer data-testid="question-action-buttons-container">
+      <Tooltip tooltip={bookmarkTooltip}>
+        <Button
+          onlyIcon
+          icon="bookmark"
+          iconSize={ICON_SIZE}
+          onClick={handleBookmark}
+          color={bookmarkButtonColor}
+        />
+      </Tooltip>
+
+      <TippyPopoverWithTrigger
+        key="extra-actions-menu"
+        placement="bottom-end"
+        renderTrigger={({ onClick }) => (
+          <Button
+            onClick={onClick}
+            onlyIcon
+            icon="ellipsis"
+            iconSize={ICON_SIZE}
+          />
+        )}
+        popoverContent={
+          <PopoverContainer>
+            <div>
+              <PLUGIN_MODERATION.QuestionModerationButton
+                question={question}
+                VerifyButton={Button}
+                verifyButtonProps={{
+                  iconSize: ICON_SIZE,
+                  borderless: true,
+                }}
+              />
+            </div>
+            {isDataset && (
+              <div>
+                <Button
+                  icon="notebook"
+                  iconSize={ICON_SIZE}
+                  borderless
+                  onClick={() => {
+                    setQueryBuilderMode("dataset", {
+                      datasetEditorTab: "query",
+                    });
+                  }}
```

Comment:
Should we maybe extract the click handlers with `useCallback` hooks to make this a bit more readable and less nested?